File: 2_KDTree/src/KDTree.cpp
#include <iostream>
#include <memory>
#include <thread>
#include <random>
#include <cstdlib>

#include <Eigen/Dense>

#include "open3d/Open3D.h"


int main(int argc, char *argv[]) 
{
    //----- Create a pointer to store the point cloud -----//
    auto cloud_ptr = std::make_shared<open3d::geometry::PointCloud>();
    open3d::visualization::Visualizer visualizer; //Create a visualizer object
    // Create a pointcloud from a file Return an empty pointcloud, if fail to read the file.
    cloud_ptr = open3d::io::CreatePointCloudFromFile(argv[1]);
    if (open3d::io::ReadPointCloud(argv[1], *cloud_ptr))
    {
        open3d::utility::LogInfo("Successfully read {}", argv[1]);
    }
    else
    {
        open3d::utility::LogInfo("Failed to read {}", argv[1]);
        return 1;
    }

    //----- Paint all points gray -----//
    open3d::utility::LogInfo("Paint all points Gray...");
    Eigen::Vector3d gray = {0.5, 0.5, 0.5}; 
    for(size_t i = 0; i < cloud_ptr->points_.size(); i++){
        cloud_ptr -> colors_[i] = gray;
    }
    //----- Build a KDTree -----//
    open3d::utility::LogInfo("Build KDTree...");
    open3d::geometry::KDTreeFlann kdtree;
    kdtree.SetGeometry(*cloud_ptr);

    // Pick the 1500th points as the anchor point and paint it red
    open3d::utility::LogInfo("Paint the 1500th point red.");
    cloud_ptr -> colors_[1500] = Eigen::Vector3d(1, 0, 0);


    //----- K Nearest Neighbors Search -----//
    open3d::utility::LogInfo("Find its 200 nearest neighbors, and paint them blue.");
    int knn = 200;
    std::vector<int> new_indices_vec(knn);
    std::vector<double> new_dists_vec(knn);
    kdtree.SearchKNN(cloud_ptr->points_[1500], knn, new_indices_vec, new_dists_vec);
    for (size_t i = 1; i < new_indices_vec.size(); i++) {
        //open3d::utility::LogInfo("{:d}, {:f}", (int)new_indices_vec[i], sqrt(new_dists_vec[i]));
        cloud_ptr -> colors_[new_indices_vec[i]] = Eigen::Vector3d(0, 0, 1);
    } // Skip the first index since it is the anchor point itself.


    //----- Radius Nearest Neighbors Search -----//
    //query all points with distances to the anchor point less than a given radius. 
    //paint these points with a green color.
    double radius = 0.2;
    std::vector<int> indices_vec(knn);
    std::vector<double> dists_vec(knn);
    kdtree.SearchRadius(cloud_ptr->points_[1500], radius, indices_vec, dists_vec);
    for (size_t i = 1; i < indices_vec.size(); i++) {
        cloud_ptr -> colors_[indices_vec[i]] = Eigen::Vector3d(0, 1, 0);
    } // Skip the first index since it is the anchor point itself.  


    //----- Hybrid search -----//
    // Radius K Neasrest Neighbors Search (Hybrid search)
    std::vector<int> hybrid_indices_vec(knn);
    std::vector<double> hybrid_dists_vec(knn);
    kdtree.SearchHybrid(cloud_ptr->points_[1500], radius, 50, hybrid_indices_vec, hybrid_dists_vec);
    for (size_t i = 1; i < hybrid_indices_vec.size(); i++) {
        cloud_ptr -> colors_[hybrid_indices_vec[i]] = Eigen::Vector3d(0, 1, 0.8);
    } // Skip the first index since it is the anchor point itself.  


    //----- Point cloud Visualization -----//
    visualizer.CreateVisualizerWindow("Open3D Down Sampling", 1600, 900);
    visualizer.AddGeometry(cloud_ptr);
    visualizer.Run();
    visualizer.DestroyVisualizerWindow();
    return 0;
}

File: 2_KDTree/KDTree.py
import open3d as o3d
import numpy as np
import matplotlib.pyplot as plt
import copy
import os
import sys

import open3d_tutorial as o3dtut

##----- Build KDTree from point cloud -----##
print("Testing KDTree in Open3D ...")
print("Load a point cloud and paint it gray")
pcd = o3d.io.read_point_cloud("../test_data/Feature/cloud_bin_0.pcd")
pcd.paint_uniform_color([0.5, 0.5, 0.5])
pcd_tree = o3d.geometry.KDTreeFlann(pcd)

##----- Find neighbouring points -----##
# pick the 1500th point as the anchor point and paint it red.
print("Paint the 1500th point red.")
pcd.colors[1500] = [1, 0, 0]

##----- Using search_knn_vector_3d -----##
# returns a list of indices of the k nearest neighbors of the anchor point.
# These neighboring points are painted with blue color. 
# skip the first index since it is the anchor point itself.
print("Find its 200 nearest neighbors, and paint them blue.")
[k, idx, _] = pcd_tree.search_knn_vector_3d(pcd.points[1500], 200)
np.asarray(pcd.colors)[idx[1:], :] = [0, 0, 1]

##----- Using search_radius_vector_3d -----##
# to query all points with distances to the anchor point less than a given radius.
# paint these points with a green color.
print("Find its neighbors with distance less than 0.2, and paint them green.")
[k, idx, _] = pcd_tree.search_radius_vector_3d(pcd.points[1500], 0.2)
np.asarray(pcd.colors)[idx[1:], :] = [0, 1, 0]

##----- Using search_hybrid_vector_3d -----##
# Radius K Neasrest Neighbors Search (Hybrid search)
[k, idx, _] = pcd_tree.search_hybrid_vector_3d(pcd.points[1500], 0.2, 50)
np.asarray(pcd.colors)[idx[1:], :] = [0, 1, 0.8]

print("Visualize the point cloud.")
o3d.visualization.draw_geometries([pcd],
                                  zoom=0.5599,
                                  front=[-0.4958, 0.8229, 0.2773],
                                  lookat=[2.1126, 1.0163, -1.8543],
                                  up=[0.1007, -0.2626, 0.9596],
                                  point_show_normal=False)